rusk/src/lib/http.rs

```diff
@@ -0,0 +1,295 @@
+                    }
+                    // The stream has stopped producing messages, and we
+                    // should close it and stop. The client likely has done
+                    // this on purpose, and it's a part of the normal
+                    // operation of the server.
+                    None => {
+                        let _ = stream.close(Some(CloseFrame {
+                            code: CloseCode::Normal,
+                            reason: Cow::from("Stream stopped"),
+                        })).await;
+                        break;
+                    }
+                };
+                match req {
+                    // We received a valid request and should spawn a new task to handle it
+                    Ok(req) => {
+                        task::spawn(handle_execution(
+                            sources.clone(),
+                            req,
+                            responder.clone(),
+                        ));
+                    },
+                    Err(e) => {
+                        let _ = stream.close(Some(CloseFrame {
+                            code: CloseCode::Error,
+                            reason: Cow::from(e.to_string()),
+                        })).await;
+                        break;
+                    }
+                }
+
+            }
+        }
+    }
+}
+
+struct ExecutionService {
+    sources: Arc<DataSources>,
+    shutdown: broadcast::Receiver<Infallible>,
+}
+
+impl Service<Request<Body>> for ExecutionService
+// where
+//     Q: QueryRaw,
+{
+    type Response = Response<Body>;
+    type Error = ExecutionError;
+    type Future = Pin<
+        Box<
+            dyn Future<Output = Result<Self::Response, Self::Error>>
+                + Send
+                + 'static,
+        >,
+    >;
+
+    fn poll_ready(
+        &mut self,
+        _: &mut Context<'_>,
+    ) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    /// Handle the HTTP request.
+    ///
+    /// A request may be a "normal" request, or a WebSocket upgrade request. In
+    /// the former case, the request is handled on the spot, while in the
+    /// latter task running the stream handler loop is spawned.
+    fn call(&mut self, mut req: Request<Body>) -> Self::Future {
+        let sources = self.sources.clone();
+        let shutdown = self.shutdown.resubscribe();
+
+        Box::pin(async move {
+            if hyper_tungstenite::is_upgrade_request(&req) {
+                let (response, websocket) =
+                    hyper_tungstenite::upgrade(&mut req, None)?;
+
+                task::spawn(handle_stream(sources, websocket, shutdown));
+
+                Ok(response)
+            } else {
+                let (_, req_body) = req.into_parts();
+
+                let body = body::to_bytes(req_body).await?;
+                let execution_request = serde_json::from_slice(&body)?;
```

Comment:
Not a feature request, but just to ensure we thought about this: we only support text via HTTP?

Reply:
Nice point, going to add